core/rawdb/chain_iterator.go

```diff
@@ -187,41 +194,109 @@ func InitDatabaseFromFreezer(db ethdb.Database) {
 // We can write tx index tail flag periodically even without the whole indexing
 // procedure is finished. So that we can resume indexing procedure next time quickly.
 func IndexTransactions(db ethdb.Database, from uint64, to uint64) {
-	// hashTxs calculates transaction hash in advance using the multi-routine's
-	// concurrent computing power.
-	hashTxs := func(block *types.Block) {
-		for _, tx := range block.Transactions() {
-			tx.Hash()
-		}
+	// short circuit for invalid range
+	if from >= to {
+		return
 	}
-	// writeIndices injects txlookup indices into the database.
-	writeIndices := func(batch ethdb.Batch, block *types.Block) {
-		WriteTxLookupEntries(batch, block)
-		if block.NumberU64() == to-1 || block.NumberU64()%10000 == 0 {
-			WriteTxIndexTail(batch, block.NumberU64())
+	var (
+		hashesCh, abortCh = iterateTransactions(db, from, to, true)
+		batch             = db.NewBatch()
+		start             = time.Now()
+		logged            = start.Add(-7 * time.Second)
+		lastNum           uint64
```

Comment:
Shouldn't `lastNum` be initialized?

Reply:
Not necessarily. If we don't write anything, it will be zero. However, you are right that if that _does_ happen (the chans are aborted for whatever reason), we shouldn't write an index tail of `0` to db..

Reply:
maybe
```
lastNum           = *ReadTxIndexTail(db)
```
?